### PR TITLE
Fix for Colored String Case

### PR DESCRIPTION
**Pull request by hotpocket184**
### Description
This PR fixes an issue when using the string case expression with colored strings. Not exactly sure why it's showing commits from a different branch, sorry about that.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/4602

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 1]
```diff
@@ -209,5 +219,23 @@ private static String toKebabCase(String str, int mode) {
...
+		}
+		matcher = Pattern.compile("[&§][0-9a-fk-orx]").matcher(colored);
+		i = 0;
+		while (matcher.find()) {
+			stringBuilder.insert(colored.indexOf(matcher.group()) + i, matcher.group());
+			colored = colored.replaceFirst("[&§][0-9a-fk-orx]", "");
+			i = i + 2;
+		}
+		return stringBuilder.toString();
+	}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java:240]:
```suggestion
	}

```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Does this only work for hex and color tags? We just probably support any sort of tag contained within `<>`

Also, you should update https://github.com/SkriptLang/Skript/blob/master/src/test/skript/tests/syntaxes/expressions/ExprStringCase.sk to test this

**Comment by AyhamAl-Ali** [comment]
# Proposal
I had an idea about this, probably easier and cleaner here how it goes:

- You make a little parser ([something like this](https://github.com/SkriptLang/Skript/blob/63ab42eb7f36cf4e6bf8dd54220fbfb4fcf0992d/src/main/java/ch/njol/skript/doc/Documentation.java#L175-L219)) to walk each character from last to beginning (important) and find chat colors/tags/hex codes by:
- Read first > and go to next < then regex check the inside of it
- Read & and chacter on its right and do regex heck
- Read # and the 3-6 characters on its right and do regex heck

Then cut them from the text, save it with its first character position (most left character) in a map and move on

Once the text is filtered, to restore it you do this:
- Inverse the map (related to important above)
- Loop the map then insert each block at its position in the filtered text

If you think you can do that and it's easier go ahead otherwise it's just an idea

# Notes
If you wanted to keep your current code, make sure to compile all regex patterns outside the methods and assign them to their own fields

**Comment by TheLimeGlass** [comment]
There is `ch.njol.skript.util.chat.ChatMessages#stripStyles(String)` for removing the colour formats, it may also be better to put your utility methods inside `ch.njol.skript.util.chat.ChatMessages`

**Comment by hotpocket184** [comment]
> There is `ch.njol.skript.util.chat.ChatMessages#stripStyles(String)` for removing the colour formats, it may also be better to put your utility methods inside `ch.njol.skript.util.chat.ChatMessages`

The only problem with the strip styles is that it throws an index out of bounds exception. I will add it to `ch.njol.skript.util.chat.ChatMessages` though, thanks for the suggestion.

File: src/main/java/ch/njol/skript/expressions/ExprStringCase.java [hunk 2]
```diff
@@ -126,9 +124,9 @@ protected String[] get(Event e) {
 						break;
 					case 1: // Proper Case
 						String cStr = strs[i];
-						strs[i] = strs[i].replaceAll("[&§][0-9a-fk-orx]", "").replaceAll("<#(?:[0-9a-fA-F]{3}){1,2}$>", "");
+						strs[i] = strs[i].replaceAll("[&§][0-9a-fk-orx]", "").replaceAll("(<?<=<)(.*?)(?=>)", "");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprStringCase.java:127]: can use `cStr` instead of `strs[i]` here

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
I think a test would be good for this

**Pull request closed without merging**